lib/ansible/utils/encrypt.py

```diff
@@ -33,31 +40,140 @@
+        return settings.get('salt') or random_salt(salt_size)
+
+    def _rounds(self, **settings):
+        if settings.get('rounds') == self.algo_data.implicit_rounds:
+            # Passlib does not include the rounds if it is the same as implict_rounds.
+            # Make crypt lib behave the same, by not explicitly specifying the rounds in that case.
+            return None
+        return settings.get('rounds')
+
+    def _encrypt(self, secret, salt, rounds):
+        if rounds is None:
+            saltstring = "$%s$%s" % (self.algo_data.crypt_id, salt)
+        else:
+            saltstring = "$%s$rounds=%d$%s" % (self.algo_data.crypt_id, rounds, salt)
+        result = crypt.crypt(secret, saltstring)
+
+        # crypt.crypt returns None if it cannot parse saltstring
+        # None as result would be interpreted by the some modules (user module)
+        # as no password at all.
+        if not result:
+            raise self.error_type("crypt.crypt does not support '%s' algorithm" % self.algorithm)
+
+        return result
+
+
+class PasslibHash(BaseHash):
+    def __init__(self, algorithm, error_type):
+        super(PasslibHash, self).__init__(algorithm, error_type)
+
+        if not PASSLIB_AVAILABLE:
+            raise error_type("passlib must be installed to hash with '%s'" % algorithm)
+
         try:
-            crypt = getattr(passlib.hash, encrypt)
+            self.crypt_algo = getattr(passlib.hash, algorithm)
         except:
-            raise AnsibleError("passlib does not support '%s' algorithm" % encrypt)
+            raise error_type("passlib does not support '%s' algorithm" % algorithm)
+
+    def hash(self, secret, **settings):
+        self._clean_salt(settings)
+        self._clean_rounds(settings)
+        return self._encrypt(secret, settings)
 
-        if salt_size:
-            result = crypt.encrypt(result, salt_size=salt_size)
-        elif salt:
-            if crypt._salt_is_bytes:
-                salt = to_bytes(salt, encoding='ascii', errors='strict')
+    def _clean_salt(self, settings):
+        if "salt" in settings:
+            if issubclass(self.crypt_algo, HasRawSalt):
+                settings["salt"] = to_bytes(settings["salt"], encoding='ascii', errors='strict')
             else:
-                salt = to_text(salt, encoding='ascii', errors='strict')
-            result = crypt.encrypt(result, salt=salt)
+                settings["salt"] = to_text(settings["salt"], encoding='ascii', errors='strict')
+
+    def _clean_rounds(self, settings):
+        algo_data = self.algorithms.get(self.algorithm)
+        if algo_data and algo_data.implicit_rounds and "rounds" not in settings:
+            # Ensure passlib behaves the same as crypt in case no rounds were specified.
+            # That means instead of using the passlib default for the rounds the crypt default is used.
+            settings["rounds"] = algo_data.implicit_rounds
```

Comment:
For _clean_salt and _clean_rounds, just pass in the one setting you care about and then return the modified version of it.  Change the container in the caller.  functional programming like that is easier to test and easier to reason about.